[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56718 [WIP] Added skeleton of python dispatch key
* **#57426 [nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning**

Differential Revision: [D28293191](https://our.internmc.facebook.com/intern/diff/D28293191)

Chillee committed:
[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning

[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

ZolotukhinM commented on torch/lib/pyston3.8/site-packages/caffe2/contrib/aten/aten_test.py:
```diff
@@ -0,0 +1,136 @@
+
+
+
+
```
Ouch, rebase went wrong?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2426,9 +2423,59 @@ Tensor* tensorexpr::computeOperandValue(
+          outputShape,
+          outputType);
+    }
+    case aten::transpose: {
```
Can we add some correctness tests for these? We could write them without relying on the fuser, we can feed graphs directly to `TensorExprKernel` like it is done, e.g. here:
https://github.com/pytorch/pytorch/blob/b5b158a6c6de94dfb983b447fa33fea062358844/test/test_tensorexpr_pybind.py#L125-L134

ZolotukhinM reviewed (commented):
Generally looks good. I'd prefer to have some tests with it and also it seems that some unrelated stuff has creeped in

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

ZolotukhinM commented on test/test_tensorexpr_pybind.py:
```diff
@@ -206,5 +206,104 @@ def forward(self, x, y):
+        res1 = kernel.run((x,))
+        res2 = kernel.fallback((x,))
+        correct = f(x)
+        np.testing.assert_allclose(res1.numpy(), correct.numpy(), atol=2e-3)
```
Nit: I know we use this `2e-3` accuracy in some existing tests, but I think we really should be more accurate than this :) Can we change it to `1e-5`?

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/external_functions.cpp:
```diff
@@ -225,6 +225,24 @@ void nnc_aten_addmm(
   }
 }
 
+void nnc_aten_digamma(
```
It would be good to add a test for this too, like we have for other external calls (in https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_external_calls.cpp)

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -3223,4 +3276,4 @@ void TensorExprKernel::runFast(
   // Call the kernel.
   codegen_->call_raw(args);
-}
+}
```
Nit: unnecessary whitespace edit.

ZolotukhinM reviewed (approved):
Looks good!

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"


Differential Revision: [D28293191](https://our.internmc.facebook.com/intern/diff/D28293191)

[ghstack-poisoned]

Chillee committed:
Update on "[nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning"


Differential Revision: [D28293191](https://our.internmc.facebook.com/intern/diff/D28293191)

[ghstack-poisoned]

Chillee commented:
@chillee has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28293191).
